File: asm/src/main/java/dev/openrs2/asm/InsnListUtils.java
package dev.openrs2.asm;

import java.util.ArrayList;

import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;

public final class InsnListUtils {
	public static boolean replaceSimpleExpression(InsnList list, AbstractInsnNode last, AbstractInsnNode replacement) {
		var deadInsns = new ArrayList<AbstractInsnNode>();

		var height = 0;
		var insn = last;
		do {
			var metadata = StackMetadataKt.stackMetadata(insn);
			if (insn != last) {
				deadInsns.add(insn);
				height -= metadata.getPushes();
			}
			height += metadata.getPops();

			if (height == 0) {
				deadInsns.forEach(list::remove);

				if (replacement != null) {
					list.set(last, replacement);
				} else {
					list.remove(last);
				}

				return true;
			}

			insn = insn.getPrevious();
		} while (insn != null && insn.getType() != AbstractInsnNode.LABEL && InsnNodeUtils.isPure(insn));

		return false;
	}

	public static boolean deleteSimpleExpression(InsnList list, AbstractInsnNode last) {
		return replaceSimpleExpression(list, last, null);
	}

	private InsnListUtils() {
		/* empty */
	}
}

File: asm/src/main/java/dev/openrs2/asm/InsnListUtils.kt
package dev.openrs2.asm

import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.InsnList

private val ANY_INSN = { _: AbstractInsnNode -> true }

public fun getExpression(
    last: AbstractInsnNode,
    filter: (AbstractInsnNode) -> Boolean = ANY_INSN
): List<AbstractInsnNode>? {
    val expr = mutableListOf<AbstractInsnNode>()

    var height = 0
    var insn: AbstractInsnNode? = last
    do {
        val (pops, pushes) = insn!!.stackMetadata
        if (insn !== last) {
            expr.add(insn)
            height -= pushes
        }
        height += pops

        if (height == 0) {
            return expr.asReversed()
        }

        insn = insn.previous
    } while (insn != null && insn.isSequential && filter(insn))

    return null
}

public fun InsnList.replaceExpression(
    last: AbstractInsnNode,
    replacement: AbstractInsnNode,
    filter: (AbstractInsnNode) -> Boolean = ANY_INSN
): Boolean {
    val expr = getExpression(last, filter) ?: return false
    expr.forEach(this::remove)
    this[last] = replacement
    return true
}

public fun InsnList.deleteExpression(
    last: AbstractInsnNode,
    filter: (AbstractInsnNode) -> Boolean = ANY_INSN
): Boolean {
    val expr = getExpression(last, filter) ?: return false
    expr.forEach(this::remove)
    remove(last)
    return true
}
